### ISSUE TITLE
Select the next maps for voting based on map size and players on server

### ISSUE THREAD

**Issue by ETLegacyTeam** [etlegacy/etlegacy#1318]
Author: @alkhimey (artium)
Date: 2019-07-05
Redmine Issue: 1318, https://dev.etlegacy.com/issues/1318

---

I hope i am not mistaken and this is part of the mod and not something external. At the end of each round there is a map voting section.
Usually a server has a large number of maps and at the end of each round a subset is selected to be voted for the next map based on some algorithm.
My suggestion is to choose preferably maps that align well with the number of players in the server.

I have seen many times servers emptying because a map too small for the number of players was voted and on the other hand, playing Capuzzo airport 6 on 6 is really not fun.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2019-07-05T06:29:43Z

---

This is indeed part of of the mod. Your suggestion sounds like a good improvement, but we however don't have access to map size information (apart the few that have it somehow in the name).
I'd say this is more a server administration issue: admins should ensure map selection is adequate for their audience. I don't think we'd be able to find a good technical solution given the current limitation.

**Comment by ETLegacyTeam** [comment]
From: @alkhimey (artium)
Date: 2019-07-05T11:27:21Z

---

Spyhawk wrote:
> This is indeed part of of the mod. Your suggestion sounds like a good improvement, but we however don't have access to map size information (apart the few that have it somehow in the name).
> I'd say this is more a server administration issue: admins should ensure map selection is adequate for their audience. I don't think we'd be able to find a good technical solution given the current limitation.

Usually, the number of players varies through the day/week/year so admins can't really know which map sizes to put there in advance.

You could add an option for the admin to provide the map size information in a script, and if this information available, use it for this feature.

**Comment by ETLegacyTeam** [comment]
From: @rmarquis (Spyhawk)
Date: 2019-07-06T09:12:58Z

---

I don't think the work required is worth it tbh. Requiring admins to fill data for dozens or hundreds of maps simply won't work. Admins should monitor adjust their map selection to their average audience, and fill the servers with a min amount of bots during lull time. Easy workaround.

I'm tempted to close this ticket as 'invalid'/'won't do', if nobody objects to it in the next few days.

**Comment by ETLegacyTeam** [comment]
From: @isRyven (ryven)
Date: 2019-07-08T08:45:35Z

---

If we could manipulate map-pool from the lua by implementing corresponding API and hooks, then the functionality could be implemented outside the C-world (using lua) as an external module, to avoid code base clutter. Any volunteer could then fill the generic meta information about the maps in data structure format (i suggest JSON, so it could be reused in different environments), hosted somewhere on github, which could be used to generate the proper map-pool. 

Would also be cool if we could manipulate the campaigns this way by, again, extending lua API, and implement our dynamic-campaign feature in lua, which could also rely on map meta information to compose best possible map combination (like, there are maps that are split onto several pieces, which could all be added in a single continuous campaign).

The proper api, hooks and map meta structure would need to be defined to handle all required cases. Perhaps map meta information could also be used to print objective information in shoutcaster mode - if required fields are set (which would be simple enough to do for the competitive map-pool as it's not so big).

However, this is just an idea.

**Comment by oheikk** [comment]
The current implementation counts bots as well as spectators as "players". This leads to the situation where, if there are e.g. 10 bots in the team and 10 players spectating hoping for the opportunity to vote a smaller map so they can join a team and replace the bots, they'll only ever see maps for 20+ players. (Extreme example for the sake of clarity.)

Nobody cares what bots want, but spectators might want to vote for a map and join if they get their wish. It would be better to only count real clients, including spectators.

**Comment by yaku13** [comment]
I agree to a comment above. With bots on server mapevoteplayerscount.cfg has not much sense. Would be nice to at least allow to exclude them from players count.

**Comment by Aranud** [comment]
> The current implementation counts bots as well as spectators as "players". This leads to the situation where, if there are e.g. 10 bots in the team and 10 players spectating hoping for the opportunity to vote a smaller map so they can join a team and replace the bots, they'll only ever see maps for 20+ players. (Extreme example for the sake of clarity.)
> 
> Nobody cares what bots want, but spectators might want to vote for a map and join if they get their wish. It would be better to only count real clients, including spectators.

Fixed
